File: src/main/java/com/github/j5ik2o/fecruircbot/CrucibleEventListener.java
package com.github.j5ik2o.fecruircbot;

import java.io.IOException;
import java.util.List;

import org.jibble.pircbot.IrcException;
import org.jibble.pircbot.NickAlreadyInUseException;
import org.jibble.pircbot.PircBot;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.DisposableBean;
import org.springframework.beans.factory.InitializingBean;

import com.atlassian.crucible.event.ReviewCommentEvent;
import com.atlassian.crucible.event.ReviewCreatedEvent;
import com.atlassian.crucible.event.ReviewStateChangedEvent;
import com.atlassian.crucible.event.ReviewUpdatedEvent;
import com.atlassian.crucible.spi.PermId;
import com.atlassian.crucible.spi.data.DetailedReviewData;
import com.atlassian.crucible.spi.data.ProjectData;
import com.atlassian.crucible.spi.data.ReviewData;
import com.atlassian.crucible.spi.services.ProjectService;
import com.atlassian.crucible.spi.services.ReviewService;
import com.atlassian.event.api.EventListener;
import com.atlassian.event.api.EventPublisher;
import com.atlassian.sal.api.ApplicationProperties;
import com.atlassian.sal.api.pluginsettings.PluginSettings;
import com.atlassian.sal.api.pluginsettings.PluginSettingsFactory;

public class CrucibleEventListener implements DisposableBean,
    InitializingBean {
  private static final Logger LOGGER = LoggerFactory
      .getLogger(CrucibleEventListener.class);
  private static final Logger log = LoggerFactory
      .getLogger(CrucibleEventListener.class);
  private final EventPublisher eventPublisher;
  private final ApplicationProperties applicationProperties;
  private final ReviewService reviewService;
  private final ProjectService projectService;

  private final PircBot pircBot = new PircBot() {
    {
      setName("cru-irc-bot");
    }
  };
  private PluginSettingsFactory pluginSettingsFactory;

  public CrucibleEventListener(EventPublisher eventPublisher,
      ApplicationProperties applicationProperties,
      ReviewService reviewService,
      ProjectService projectService,
      PluginSettingsFactory pluginSettingsFactory) {
    this.eventPublisher = eventPublisher;
    this.applicationProperties = applicationProperties;
    this.reviewService = reviewService;
    this.projectService = projectService;
    this.pluginSettingsFactory = pluginSettingsFactory;
    eventPublisher.register(this);
  }

  private String getChannelName(PluginSettings settings,
      String projectId) {
    return (String) settings.get(IrcBotChannelConfig.class
        .getName() + "_" + projectId + ".channelName");
  }

  private boolean isIrcBotChannelEnable(
      PluginSettings settings, String projectId) {
    return Boolean.parseBoolean((String) settings
        .get(IrcBotChannelConfig.class.getName() + "_"
            + projectId + ".enable"));
  }

  private boolean isIrcBotChannelNotice(
      PluginSettings settings, String projectId) {
    return Boolean.parseBoolean((String) settings
        .get(IrcBotChannelConfig.class.getName() + "_"
            + projectId + ".notice"));
  }

  private boolean isIrcBotEnable(PluginSettings settings) {
    boolean enable =
        Boolean.parseBoolean((String) settings
            .get(IrcBotGlobalConfig.class.getName()
                + ".enable"));
    return enable;
  }

  private String getIrcServerName(PluginSettings settings) {
    return (String) settings.get(IrcBotGlobalConfig.class
        .getName() + ".ircServerName");
  }

  private Integer getIrcServerPort(PluginSettings settings) {
    String ircServerPort =
        (String) settings.get(IrcBotGlobalConfig.class
            .getName() + ".ircServerPort");
    if (ircServerPort != null) {
      return Integer.parseInt(ircServerPort);
    }
    return null;
  }

  @EventListener
  public void onReviewCreate(ReviewCreatedEvent event) {
    PermId<ReviewData> reviewId = event.getReviewId();
    sendNotification(reviewId, "created");
  }

  private void autoConnect(PluginSettings settings)
      throws NickAlreadyInUseException, IOException,
      IrcException {
    if (pircBot.isConnected()) {
      return;
    }

    String ircServerName = getIrcServerName(settings);
    LOGGER.debug("irc server name = " + ircServerName);
    Integer ircServerPort = getIrcServerPort(settings);
    LOGGER.debug("irc server port = " + ircServerPort);
    if (ircServerPort != null
        && ircServerPort.intValue() != 0) {
      pircBot
          .connect(ircServerName, ircServerPort.intValue());
    } else {
      pircBot.connect(ircServerName);
    }

    // List<Project> projects = projectManager.getProjectObjects();
    // for (Project project : projects) {
    // String projectId = project.getId().toString();
    // String projectName = project.getName();
    // LOGGER.debug(String.format("projectName = %s, projectId = %s",
    // projectName, projectId));
    // if (isIrcBotEnable(settings)
    // && isIrcBotChannelEnable(settings, projectId)) {
    // String channelName = getChannelName(settings, projectId);
    // LOGGER.debug(String.format("channelName = %s", channelName));
    // pircBot.joinChannel(channelName);
    // }
    // }
  }

  private void sendNotification(PermId<ReviewData> reviewId,
      String message) {

    PluginSettings settings =
        pluginSettingsFactory.createGlobalSettings();

    DetailedReviewData review =
        this.reviewService.getReviewDetails(reviewId);
    ProjectData project =
        this.projectService
            .getProject(review.getProjectKey());

    String url =
        applicationProperties.getBaseUrl() + "/cru/"
            + reviewId.getId();
    String projectUrl =
        applicationProperties.getBaseUrl() + "/project/"
            + project.getKey();

    if (isIrcBotEnable(settings) == false
        || isIrcBotChannelEnable(settings, project.getKey()) == false) {
      return;
    }

    try {
      autoConnect(settings);
      String channelName =
          getChannelName(settings, project.getKey());
      pircBot.joinChannel(channelName);

      String msg =
          "Crucible review <a href=\"" + url + "\"><b>"
              + reviewId.getId() + "</b></a> " + message
              + " in Project <a href=\"" + projectUrl + "\">"
              + project.getName() + "</a>";

      pircBot.sendMessage(channelName, message);
    } catch (NickAlreadyInUseException e) {
      // TODO Auto-generated catch block
      e.printStackTrace();
    } catch (IOException e) {
      // TODO Auto-generated catch block
      e.printStackTrace();
    } catch (IrcException e) {
      // TODO Auto-generated catch block
      e.printStackTrace();
    }

  }

  @EventListener
  public void onReviewUpdate(ReviewUpdatedEvent event) {
    PermId<ReviewData> reviewId = event.getReviewId();
    sendNotification(reviewId, "updated");
  }

  @EventListener
  public void onReviewComment(ReviewCommentEvent event) {
    PermId<ReviewData> reviewId = event.getReviewId();
    sendNotification(reviewId, "commeted");
  }

  @EventListener
  public void onReviewStateChange(
      ReviewStateChangedEvent event) {
    PermId<ReviewData> reviewId = event.getReviewId();

    sendNotification(reviewId, "status moved from "
        + event.getOldState().name() + " to "
        + event.getNewState().name());
  }

  @Override
  public void destroy() throws Exception {
    log.debug("Unregister review event listener");
    eventPublisher.unregister(this);
  }

  @Override
  public void afterPropertiesSet() throws Exception {
    log.debug("Register review event listener");
    eventPublisher.register(this);
  }

}

File: src/main/scala/com/github/j5ik2o/fecruircbot/CrucibleEventListener.scala
package com.github.j5ik2o.fecruircbot

import org.springframework.beans.factory.DisposableBean
import org.springframework.beans.factory.InitializingBean
import com.atlassian.event.api.EventListener
import com.atlassian.event.api.EventPublisher
import com.atlassian.sal.api.ApplicationProperties
import com.atlassian.crucible.event._
import parser._
import scala.collection.JavaConverters._
import com.atlassian.crucible.spi.PermId
import com.atlassian.crucible.spi.data._
import scala.util.control.Exception._
import com.atlassian.crucible.spi.services.{NotFoundException, ReviewService, ProjectService}
import scala._
import scala.Predef._

class CrucibleEventListener
(
  eventPublisher: EventPublisher,
  applicationProperties: ApplicationProperties,
  reviewService: ReviewService,
  protected val projectService: ProjectService,
  //protected val pluginSettingsFactory: PluginSettingsFactory,
  protected val ircBotGlobalConfigRepository: IrcBotGlobalConfigRepository,
  protected val ircBotProjectChannelConfigRepository: IrcBotProjectChannelConfigRepository
  )
  extends DisposableBean with InitializingBean with IrcConfigAccess {

  protected val name = "cru-irc-bot"

  protected def isIrcBotChannelEnable(key: String) = {
    ircBotProjectChannelConfigRepository.resolve(key) match {
      case Some(IrcBotProjectChannelConfig(true, _, _)) => true
      case _ => false
    }
  }

  protected def getIrcBotChannelName(key: String) =
    ircBotProjectChannelConfigRepository.resolve(key).get.getChannelName()

  protected def isIrcBotChannelNotice(key: String) =
    ircBotProjectChannelConfigRepository.resolve(key).get.getNotice()

  private def getReviewUrl(reviewId: PermId[ReviewData]): String =
    applicationProperties.getBaseUrl + "/cru/" + reviewId.getId

  private def getProjectUrl(projectKey: String): String =
    applicationProperties.getBaseUrl + "/project/" + projectKey

  private def formattedUser(userData: UserData) =
    "%s(%s)".format(
      if (userData.getDisplayName != null) userData.getDisplayName else "",
      if (userData.getUserName != null) userData.getUserName else "")

  private def formattedComment(commentData: CommentData) =
    "%s: %s".format(
      formattedUser(commentData.getUser),
      commentData.getMessage
    )

  private def formattedProject(reviewData: ReviewData) =
    "[%s]".format(reviewData.getPermaId.getId)

  private def formattedCommentMessage
  (projectData: ProjectData,
   reviewData: DetailedReviewData,
   commentId: PermId[CommentData],
   title: String) = {
    val commentDataOption = catching(classOf[NotFoundException]) opt reviewService.getComment(commentId)
    commentDataOption.map {
      commentData =>
        Seq(
          "%s %s".format(formattedProject(reviewData), title),
          "CommentId: %s".format(commentId.getId),
          formattedComment(commentData),
          getReviewUrl(reviewData.getPermaId) + "#c" + commentId.getId.split(":")(1)
        )
    } getOrElse {
      Seq(
        "%s %s".format(formattedProject(reviewData), title),
        "CommentId: %s".format(commentId.getId),
        getReviewUrl(reviewData.getPermaId)
      )
    }
  }


  private def formattedReviewMessage
  (projectData: ProjectData,
   reviewData: DetailedReviewData,
   title: String) = {
    val r = List(
      Some("%s %s".format(formattedProject(reviewData), title)),
      if (reviewData.getPermaId != null)
        Some("Id: %s".format(reviewData.getPermaId.getId))
      else
        None,
      if (reviewData.getName != null && reviewData.getName.isEmpty == false)
        Some("Name: %s".format(reviewData.getName))
      else
        None,
      if (reviewData.getDescription != null && reviewData.getDescription.isEmpty == false)
        Some("Description: %s".format(reviewData.getDescription))
      else
        None,
      if (reviewData.getAuthor != null)
        Some("Author: %s".format(formattedUser(reviewData.getAuthor)))
      else
        None,
      if (reviewData.getModerator != null)
        Some("Moderator: %s".format(formattedUser(reviewData.getModerator)))
      else
        None).flatten ++
      reviewData.getReviewers.reviewer.asScala.map {
        e => Some("Reviewer: %s(%s)".format(e.getDisplayName, e.getUserName))
      }.flatten.toList ++
      List(
        if (reviewData.getDueDate != null)
          Some("Due Date: %s".format(dateFormat.format(reviewData.getDueDate)))
        else
          None,
        if (reviewData.getJiraIssueKey != null)
          Some("JIRA: %s".format(reviewData.getJiraIssueKey))
        else
          None,
        Some(getReviewUrl(reviewData.getPermaId))
      ).flatten
    r.toSeq
  }

  private def formattedReviewStateChangeMessage
  (projectData: ProjectData,
   reviewData: DetailedReviewData,
   reviewOldDataState: ReviewData.State,
   reviewNewDataState: ReviewData.State,
   title: String) =
    Seq(
      "%s %s".format(formattedProject(reviewData), title),
      "%s から %s に変更しました".format(
        reviewOldDataState.name(),
        reviewNewDataState.name()
      ),
      getReviewUrl(reviewData.getPermaId)
    )

  private def formattedReviewerCompletedMessage
  (projectData: ProjectData,
   reviewData: DetailedReviewData,
   reviewerData: ReviewerData,
   title: String) =
    Seq(
      "%s %s".format(formattedProject(reviewData), title),
      "%s(%s)はレビューを完了させました。".format(
        reviewerData.getDisplayName,
        reviewerData.getUserName
      ),
      getReviewUrl(reviewData.getPermaId)
    )

  private def formattedonReviewItemRevisionDataChangeMessage
  (projectData: ProjectData,
   reviewData: DetailedReviewData,
   addedRevisions: Iterable[ReviewItemRevisionData],
   removedRevisions: Iterable[ReviewItemRevisionData],
   title: String): Seq[String] = {

    def buildMessages(header: String, it: Iterable[ReviewItemRevisionData]) = {
      it.toList match {
        case Nil => None
        case s => Some(
          header :: s.map {
            e => "r%s, %s, %s".format(
              e.getRevision,
              dateFormat.format(e.getAddDate),
              e.getPath)
          }
        )
      }
    }
    val result = "%s %s".format(formattedProject(reviewData), title) ::
      (buildMessages("追加リビジョン:", addedRevisions).flatten.toList ++
        buildMessages("削除リビジョン:", removedRevisions).flatten.toList)
    (result.:+(getReviewUrl(reviewData.getPermaId))).toSeq
  }

  @EventListener
  def onReviewCommentCreated(event: CommentCreatedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)
    sendMessages(
      projectData.getKey,
      formattedCommentMessage(
        projectData,
        reviewData,
        event.getCommentId,
        "コメントが投稿されました"))
  }

  @EventListener
  def onReviewCommentDeleted(event: CommentDeletedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)
    sendMessages(
      projectData.getKey,
      formattedCommentMessage(
        projectData,
        reviewData,
        event.getCommentId,
        "コメントが削除されました"))
  }


  @EventListener
  def onReviewCommentUpdated(event: CommentUpdatedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)
    sendMessages(
      projectData.getKey,
      formattedCommentMessage(
        projectData,
        reviewData,
        event.getCommentId,
        "コメントが更新されました"))
  }

  @EventListener
  def onReviewStateChange(event: ReviewStateChangedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)
    sendMessages(
      projectData.getKey,
      formattedReviewStateChangeMessage(
        projectData, reviewData,
        event.getOldState,
        event.getNewState,
        "レビューの状態が変更されました"))
  }

  @EventListener
  def onReviewCreated(event: ReviewCreatedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)
    sendMessages(
      projectData.getKey,
      formattedReviewMessage(
        projectData,
        reviewData,
        "レビューが作成されました"))
  }

  @EventListener
  def onReviewUpdated(event: ReviewUpdatedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)
    sendMessages(
      projectData.getKey,
      formattedReviewMessage(
        projectData,
        reviewData,
        "レビューが更新されました"))
  }

  @EventListener
  def onReviewDeleted(event: ReviewDeletedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)
    sendMessages(
      projectData.getKey,
      formattedReviewMessage(
        projectData,
        reviewData,
        "レビューが削除されました"))
  }

  @EventListener
  def onReviewerCompleted(event: ReviewerCompletedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)
    sendMessages(
      projectData.getKey,
      formattedReviewerCompletedMessage(
        projectData,
        reviewData,
        event.getReviewer,
        "レビューアがレビューを完了にしました"))
  }

  @EventListener
  def onReviewerUncompleted(event: ReviewerUncompletedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)
    sendMessages(
      projectData.getKey,
      formattedReviewerCompletedMessage(
        projectData,
        reviewData,
        event.getReviewer,
        "レビューアがレビューを未完了にしました"))
  }

  @EventListener
  def onReviewItemRevisionDataChange(event: ReviewItemRevisionDataChangedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)

    import scala.collection.JavaConverters._

    val addedRevisionsAsJava = event.getAddedRevisions
    val removedRevisionsAsJava = event.getRemovedRevisions
    sendMessages(
      projectData.getKey,
      formattedonReviewItemRevisionDataChangeMessage(
        projectData,
        reviewData,
        if (addedRevisionsAsJava != null) addedRevisionsAsJava.asScala else Iterable.empty,
        if (removedRevisionsAsJava != null) removedRevisionsAsJava.asScala else Iterable.empty,
        "リビジョンが追加もしくは削除されました"))

  }

  @EventListener
  def onAllReviewersCompleted(event: AllReviewersCompletedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)
    sendMessages(
      projectData.getKey,
      formattedReviewMessage(
        projectData,
        reviewData,
        "すべてのレビューアがレビューを完了させました"))
  }

  @EventListener
  def onAllReviewersNoLongerCompleted(event: AllReviewersNoLongerCompletedEvent) {
    val reviewData = reviewService.getReviewDetails(event.getReviewId)
    val projectData = projectService.getProject(reviewData.getProjectKey)
    sendMessages(
      projectData.getKey,
      formattedReviewMessage(
        projectData,
        reviewData,
        "すべてのレビューアがレビューを久しく完了していません"))
  }

  def listReview(asc: Boolean) = {
    val r = reviewService.getAllReviews(true)
    if (r != null) {
      r.asScala.filter {
        e =>
          e.getState == ReviewData.State.Review
      }.map {
        e =>
          (e, reviewService.getAllRevisionComments(e.getPermaId))
      }.sortWith(
        (o1, o2) =>
          ((o1._2.size < o2._2.size) == asc)
      ).take(10).map {
        e =>
          "[%s:%d] %s , a:%s, m:%s : %s".
            format(
            e._1.getPermaId.getId,
            e._2.size,
            e._1.getName,
            formattedUser(e._1.getAuthor),
            formattedUser(e._1.getModerator),
            getReviewUrl(e._1.getPermaId)
          )
      }
    } else
      List.empty[String]
  }

  def onMessage(channel: String, sender: String, login: String, hostname: String, message: String) {
    LOGGER.info("c = %s, s = %s, l = %s, h = %s, m = %s".format(channel, sender, login, hostname, message))
    if (isIrcBotEnable && message.startsWith("cru-irc-bot")) {
      val b = new BotParsers
      val r = try {
        b.parse(message) match {
          case ListOpecode(ReviewOperand(CommentSortType(Asc))) => listReview(true)
          case ListOpecode(ReviewOperand(CommentSortType(Desc))) => listReview(false)
        }
      } catch {
        case e: BotParseException =>
          Seq(e.msg)
      }
      r.foreach {
        pircBot.sendMessage(channel, _)
      }
    }
  }

  def destroy {
    eventPublisher.unregister(this)
  }

  protected def connectAllChannel = {
    autoConnect
    val p = projectService.getAllProjects
    if (p != null) {
      val projects = p.asScala
      projects.map(_.getKey).filter(
        isEnableChannel(_)
      ).foreach {
        key =>
          val channelName = getIrcBotChannelName(key)
          pircBot.joinChannel(channelName)
      }
    }
  }

  def afterPropertiesSet {
    connectAllChannel
    eventPublisher.register(this)
  }

}